ext/session/session.c

```diff
@@ -2344,15 +2782,64 @@ static PHP_FUNCTION(session_start)
 }
 /* }}} */
 
-/* {{{ proto bool session_destroy(void)
+
+/* {{{ proto array session_info(void)
+   Return session internal data information */
+static PHP_FUNCTION(session_info)
+{
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	if (PS(session_status) != php_session_active) {
+		php_error_docref(NULL, E_WARNING, "Session is not active");
+		RETURN_FALSE;
+	}
+
+	ZEND_ASSERT(!Z_ISUNDEF(PS(internal_data)));
+	ZVAL_COPY(return_value, &PS(internal_data));
+	return;
+}
+/* }}} */
+
+
+/* {{{ proto int session_gc(void)
+   Perform session GC manually */
+static PHP_FUNCTION(session_gc)
+{
+	zend_long nrdels;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	if (PS(session_status) != php_session_active) {
+		php_error_docref(NULL, E_WARNING, "Session is not active");
+		RETURN_FALSE;
+	}
+
+	nrdels = php_session_gc();
+	if (nrdels < 0) {
+		RETURN_FALSE;
+	}
+	ZVAL_LONG(return_value, nrdels);
+}
+/* {{{ proto bool session_destroy([bool immediate])
    Destroy the current session and all data associated with it */
 static PHP_FUNCTION(session_destroy)
 {
-	if (zend_parse_parameters_none() == FAILURE) {
+	zend_bool immediate=0;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|b", &immediate) == FAILURE) {
 		return;
 	}
 
-	RETURN_BOOL(php_session_destroy() == SUCCESS);
+	if (immediate) {
+		/* Immediate destroy may cause random lost session by server side race condition. */
+		php_error_docref(NULL, E_NOTICE, "Immediate session data removal may cause random lost sessions. It is advised not to delete immediately");
```

Comment:
This is a total no-go. A notice for an intentionally chosen behavior (explicitly passing true).
Warn about it in docs, but definitely not force the user to use the @ operator.